src/widget/wsearchrelatedtracksmenu.cpp

```diff
@@ -0,0 +1,291 @@
+#include "widget/wsearchrelatedtracksmenu.h"
+
+#include "track/track.h"
+#include "util/qt.h"
+
+namespace {
+
+constexpr double kRelativeBpmRange = 0.06; // +/-6 %
+
+inline int bpmLowerBound(double bpm) {
+    return static_cast<int>(std::floor((1 - kRelativeBpmRange) * bpm));
+}
+
+inline int bpmUpperBound(double bpm) {
+    return static_cast<int>(std::ceil((1 + kRelativeBpmRange) * bpm));
+}
+
+QString extractCalendarYearNumberFromReleaseDate(
+        const QString& releaseDate) {
+    // TODO: Improve this poor calendar year number parser
+    // and extract the code
+    int skippedLeadingZeros = 0;
+    int countedLeadingDigits = 0;
+    const auto trimmed = releaseDate.trimmed();
+    // Count leading digits
+    for (int i = 0; i < trimmed.size(); ++i) {
+        if (!trimmed[i].isDigit()) {
+            break;
+        }
+        // Skip leading zeros
+        if (countedLeadingDigits == 0 && trimmed[i] == QChar('0')) {
+            ++skippedLeadingZeros;
+            continue;
+        }
+        // Count leading digits
+        ++countedLeadingDigits;
+    }
+    // Interpret the leading digits as the calendar year
+    // without any further validation
+    return trimmed.mid(skippedLeadingZeros, countedLeadingDigits);
+}
+
+} // namespace
+
+WSearchRelatedTracksMenu::WSearchRelatedTracksMenu(
+        QWidget* parent)
+        : QMenu(tr("Search related Tracks"), parent) {
+}
+
+bool WSearchRelatedTracksMenu::addTriggerSearchAction(
+        bool addSeparatorBeforeNextAction,
+        const QString& actionText,
+        QString /*!by-value-because-captured-by-lambda!*/ searchQuery) {
+    if (addSeparatorBeforeNextAction) {
+        addSeparator();
+    }
+    addAction(
+            mixxx::escapeTextPropertyWithoutShortcuts(actionText),
+            [this, searchQuery]() {
+                emit triggerSearch(searchQuery);
+            });
+    return false;
+}
+
+void WSearchRelatedTracksMenu::addActionsForTrack(
+        const Track& track) {
+    // NOTE: We have to explicitly use `QString` instead of `auto`
+    // when composing search queries using `QStringBuilder`. Otherwise
+    // string concatenation will fail at runtime!
+
+    // Mixing actions
+    bool addSeparatorBeforeNextAction = !isEmpty();
+    {
+        const auto keyText = track.getKeyText();
+        if (!keyText.isEmpty()) {
+            const auto actionText =
+                    tr("Key: Harmonic with \"%1\"").arg(keyText);
+            const QString searchQuery =
+                    QStringLiteral("~key:\"") +
+                    keyText +
+                    QChar('"');
+            addSeparatorBeforeNextAction = addTriggerSearchAction(
+                    addSeparatorBeforeNextAction,
+                    actionText,
+                    searchQuery);
+        }
+    }
+    {
```

Comment:
What is the need for these blocks? It seems that a simple blank line would be adequate to separate what are currently blocks.

Reply:
To prevent reusing the wrong local variables between menu entries unintentionally.

Reply:
This could happen easily when copy & pasting code.

Reply:
Okay. I wouldn't have written it that way, but it works how it is.